tests/Unit/Evolution/Systems/ScalarWave/Test_Constraints.cpp

```diff
@@ -0,0 +1,344 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "tests/Unit/TestingFramework.hpp"
+
+#include <array>
+#include <cstddef>
+#include <limits>
+#include <memory>
+#include <pup.h>
+#include <string>
+
+#include "DataStructures/DataBox/DataBox.hpp"
+#include "DataStructures/DataBox/Prefixes.hpp"  // IWYU pragma: keep
+#include "DataStructures/DataVector.hpp"        // IWYU pragma: keep
+#include "DataStructures/Tensor/EagerMath/DeterminantAndInverse.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "DataStructures/Variables.hpp"
+#include "Domain/CoordinateMaps/Affine.hpp"
+#include "Domain/CoordinateMaps/CoordinateMap.hpp"
+#include "Domain/CoordinateMaps/ProductMaps.hpp"
+#include "Domain/LogicalCoordinates.hpp"
+#include "Domain/Mesh.hpp"
+#include "Domain/Tags.hpp"  // IWYU pragma: keep
+#include "Evolution/Systems/ScalarWave/Constraints.hpp"
+#include "Evolution/Systems/ScalarWave/Tags.hpp"
+#include "NumericalAlgorithms/LinearOperators/PartialDerivatives.tpp"  // IWYU pragma: keep
+#include "NumericalAlgorithms/Spectral/Spectral.hpp"
+#include "PointwiseFunctions/AnalyticSolutions/WaveEquation/PlaneWave.hpp"
+#include "PointwiseFunctions/AnalyticSolutions/WaveEquation/RegularSphericalWave.hpp"
+#include "PointwiseFunctions/MathFunctions/Gaussian.hpp"
+#include "PointwiseFunctions/MathFunctions/MathFunction.hpp"
+#include "Utilities/ConstantExpressions.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/MakeWithValue.hpp"
+#include "Utilities/TMPL.hpp"
+#include "Utilities/TaggedTuple.hpp"
+#include "tests/Unit/Pypp/CheckWithRandomValues.hpp"
+#include "tests/Unit/Pypp/SetupLocalPythonEnvironment.hpp"
+
+// IWYU pragma: no_forward_declare Tensor
+// IWYU pragma: no_forward_declare Variables
+// IWYU pragma: no_forward_declare Tags::deriv
+
+namespace {
+// Test the return-by-value two-index constraint function using random values
+template <size_t Dim>
+void test_two_index_constraint_random(
+    const DataVector& used_for_size) noexcept {
+  pypp::check_with_random_values<1>(
+      static_cast<tnsr::ij<DataVector, Dim, Frame::Inertial> (*)(
+          const tnsr::ij<DataVector, Dim, Frame::Inertial>&)>(
+          &ScalarWave::two_index_constraint<Dim>),
+      "TestFunctions", "two_index_constraint", {{{-10.0, 10.0}}}, used_for_size,
+      1.0e-12);
+}
+
+// Test the return-by-reference two-index constraint
+// by comparing to a time-independent analytic solution (e.g. Kerr-Schild)
+template <typename Solution>
+void test_two_index_constraint_analytic(
+    const Solution& solution, const size_t grid_size_each_dimension,
+    const std::array<double, 3>& lower_bound,
+    const std::array<double, 3>& upper_bound,
+    const double error_tolerance) noexcept {
+  // Shorter names for tags.
+  constexpr size_t dim = 3;
+  using VariablesTags =
```

Comment:
This should be snake_case because you don’t instantiate this type

Reply:
changed